introduce new `initOpenNextCloudflareForDev` utility and make `getCloudflareContext` synchronous

dario-piotrowicz opened the pull request:
This PR introduces a new `initOpenNextCloudflareForDev` utility to add to the Next.js config file that makes the `getCloudflareContext` work in middlewares (or more generally, the edge runtime) during local development (via `next dev`)

It `getCloudflareContext` has also been converted to be synchronous

For more details see the changeset file: [`.changeset/chilly-dryers-begin.md`](https://github.com/opennextjs/opennextjs-cloudflare/pull/265/files#diff-75dff73666acc35e672da1f4ed34bef1d39ca6a3ca72ce956ed09ac978ea7ed4)
___

C3 PR: https://github.com/cloudflare/workers-sdk/pull/7903

Docs relative PR: https://github.com/opennextjs/docs/pull/57

___

fixes #137
fixes #226

dario-piotrowicz commented on examples/middleware/e2e/playwright.config.ts:
```diff
@@ -49,5 +49,7 @@ export default defineConfig({
     command: "pnpm preview:worker",
     url: "http://localhost:8774",
     reuseExistingServer: !process.env.CI,
+    // the app takes a bit to boot up for some reason, that's why we need a longer timeout here
+    timeout: 2 * 60 * 1000,
```
not too sure why this is needed now and it wasn't before 🤷

dario-piotrowicz replied:
ah.... probably because the boot up process takes a bit longer with `enableEdgeDevGetCloudflareContext`... weird the difference should be practically negligible (locally I don't notice any difference)

vicb commented on examples/middleware/e2e/base.spec.ts:
```diff
@@ -3,28 +3,28 @@ import { test, expect } from "@playwright/test";
 test("redirect", async ({ page }) => {
```
This looks like a fix for https://github.com/opennextjs/opennextjs-cloudflare/issues/255

Would you mind extracting the code in a distinct PR for easier tracking?

dario-piotrowicz replied:
yes of course: https://github.com/opennextjs/opennextjs-cloudflare/pull/267, sorry I didn't realise there was an issue for this 👍

dario-piotrowicz replied:
(once we merge #267 I can simply rebase this PR)

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,149 @@
+
+  const context = await getCloudflareContextFromWrangler();
+
+  vmModule.runInContext = (...args: [string, RuntimeContext, ...unknown[]]) => {
```
Extract this to a distinct function to add comments?
ref: https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/src/shared.ts#L22

vicb commented on examples/middleware/next.config.mjs:
```diff
@@ -1,4 +1,8 @@
 /** @type {import('next').NextConfig} */
 const nextConfig = {};
 
+enableEdgeDevGetCloudflareContext();
```
What about renaming this function to `initOpenNextCloudflare()`

I think it's easier for the user to understand as we expose lesst techincal details and we could add to this function later if needed.

dario-piotrowicz replied:
mh... my idea was that people would not have to always include this but only when they face the cloudflare context in the edge runtime issue (a problem that can happen to a subset of users and one that hopefully can go away in the future if Next.js decides to support nodejs middlewares), that's why I made the name more specific

`initOpenNextCloudflare()` sounds like something they always need to have (similar to [setupDevPlatform](https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/src/index.ts#L14) in next-on-pages)

> we could add to this function later if needed.

That's true, but again your comment makes it sound to me like moving forward we'd want to recommend all users to include this, which I am not sure about 😕

So based on the above I think I do prefer something like `enableEdgeDevGetCloudflareContext` which makes it very clear to the user what this function does and when/why is needed besides making clear(er) that this is optional

But if you strongly prefer a generic name like `initOpenNextCloudflare()` I'd be ok with using that (although that's not my preference)

dario-piotrowicz replied:
Resolving as this has been addressed, although I went with the name `initOpenNextCloudflareForDev` 😄 

I'm happy to reconsider if someone thinks that it is too verbose/long

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,149 @@
+  CfProperties extends Record<string, unknown> = IncomingRequestCfProperties,
+  Context = ExecutionContext,
+>(): Promise<CloudflareContext<CfProperties, Context>> {
+  const global = globalThis as unknown as {
```
could we add the type to `declare global` higher up in this file instead?

dario-piotrowicz replied:
I am not sure if you can extend the global declaration with a value indexed by a symbol, a bunch of googling lets me think that it's not possible 😕

vicb replied:
Too bad :(

We could either switch to a string or left as is then

vicb replied:
Another option:

```ts
const mySymbol = Symbol.for('internal.fake');
type MyCustomGlobal = typeof globalThis & {
  [mySymbol]: { loader: true}
}

(globalThis as MyCustomGlobal)[mySymbol] = { loader: true};
```

dario-piotrowicz replied:
> We could either switch to a string

I could really prefer not to do that because if we were to use a string and add it to the global declaration then that would leak in the user types, so with autocompletion etc... they would see such a field in their global scope (besides the fact that they could easily access it)

dario-piotrowicz replied:
> Another option:
> 
> ```ts
> const mySymbol = Symbol.for('internal.fake');
> type MyCustomGlobal = typeof globalThis & {
>   [mySymbol]: { loader: true}
> }
> 
> (globalThis as MyCustomGlobal)[mySymbol] = { loader: true};
> ```

I'm happy to go with that if you prefer, but to me looks more verbose than the current one, and I don't think it makes things clearer (a type defined inside a function is not super clear/readable in my opinion):
![Screenshot 2025-01-20 at 10 54 09](https://github.com/user-attachments/assets/46ab25b0-9748-4d42-9af0-75edc9cb71a2)

vicb replied:
Very good point for string.

The idea `typeof globalThis & {...}` would be to have this close to `declare global {...}` at the top of the file.
But then you would probably have a typing issue... which might be fixed with a function... so probably not worth it

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,149 @@
+  CfProperties extends Record<string, unknown> = IncomingRequestCfProperties,
+  Context = ExecutionContext,
+>(): Promise<CloudflareContext<CfProperties, Context>> {
+  const global = globalThis as unknown as {
```
same comment as above for the type

dario-piotrowicz replied:
as above, I think we agreed that it's not really that worth looking into changing this

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,149 @@
+      const cloudflareContext = await getCloudflareContextFromWrangler<CfProperties, Context>();
+      global[cloudflareContextInNextDevSymbol] = cloudflareContext;
+    } catch (e: unknown) {
+      if (e instanceof Error && e.message.includes("A dynamic import callback was not specified")) {
```
Is this to detect the edge runtime?
I think NEXT defines a `process.env.RUNTIME` (not sure about the exact name) and it could be clearer if we can use it?

dario-piotrowicz replied:
> Is this to detect the edge runtime?
> I think NEXT defines a process.env.RUNTIME (not sure about the exact name) and it could be clearer if we can use it?

Yes I used this try-catch to detect the edge runtime, since with debugging I couldn't see any global flag that indicated what runtime is being used, I did not think of checking values on process.env, good call, thanks 🙂, I've updated the code 👍

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,149 @@
+
+  const context = await getCloudflareContextFromWrangler();
+
+  vmModule.runInContext = (...args: [string, RuntimeContext, ...unknown[]]) => {
```
wrt `...args: [string, RuntimeContext, ...unknown[]]`

doesn't `runInContext` only have 3 args? In such a case it would be clearer to spell them out and avoid the `args[1]`. If it can have more, you can use ...args for the more part

dario-piotrowicz replied:
good point 👍

dario-piotrowicz replied:
code updated 👍

vicb commented:
Thought: Maybe we can `getPlatformProxy(...)` in `initOpenNextCloudflare()` so that `getCloudflareContext` could be made sync?

[This comment](https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/src/index.ts#L10) says "the function is an async one but it doesn't need to be awaited", it might be nice to add the rationale if we add the implementation to our repo.

dario-piotrowicz commented:
> Thought: Maybe we can `getPlatformProxy(...)` in `initOpenNextCloudflare()` so that `getCloudflareContext` could be made sync?

Yes, we could, this would mean that everyone would always have to call `initOpenNextCloudflare()`, I mentioned in [my comment](https://github.com/opennextjs/opennextjs-cloudflare/pull/265#discussion_r1922175232) that we don't need to go that route as not forcing users to do that might provide a better DX

but again if you strongly prefer us to go that route I'm ok with it

> [This comment](https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/src/index.ts#L10) says "the function is an async one but it doesn't need to be awaited", it might be nice to add the rationale if we add the implementation to our repo.

Yes, given the implementation of `next dev`, it takes some time from when Next.js loads the config and when it serves the app, we did see that that amount of time is enough for the proxy to be ready in time. It would require very little changes here, just removing the on-the-fly proxy generation from `getCloudflareContext` and changing the function signature to be sync

dario-piotrowicz commented:
@vicb I've addressed all the feedback, the only thing left is to decide whether we want to force this sort of thing via a `initOpenNextCloudflare()` and potentially convert `getCloudflareContext` to be sync or not

up to you, just let me know if you want to me make the change 👍

vicb commented:
> @vicb I've addressed all the feedback, the only thing left is to decide whether we want to force this sort of thing via a `initOpenNextCloudflare()` and potentially convert `getCloudflareContext` to be sync or not
> 
> up to you, just let me know if you want to me make the change 👍 (it doesn't really take much effort to do code-wise, it's just a matter of DX)

Give me some time to think about it, I'm not settled.

Some thoughts I have for now about `initOpenNextCloudflare()`

##Cons:##

`initOpenNextCloudflare()` always need to be called but we can have an error message when it's not called

##Pros##

- I think it's easier to document (i.e. "always call `initOpenNextCloudflare()`" vs " call enableEdgeDevGetCloudflareContext if you call getConflareContext from a middleware"). I think Vercel is working on the ability to run middlewares on the node runtime and we will add support for the edge runtime at some point. It will make it harder to explain
- I think it's easier for user to understand (They don't have to undestand what "EdgeDev" is)
- We could setup the cloudflare context in there so that `getCloudflareContext()` becomes sync. A lot of user are complaining about that on Discord and GH - See how it impacts us in the [KVCache](https://github.com/opennextjs/opennextjs-cloudflare/blob/0b54cf4c1f5850234dda206d75305761c17d32c8/packages/cloudflare/src/api/kvCache.ts#L144-L149)
- Calling the init on a per need basis (vs always calling `initOpenNextCloudflare()`) could introduce unrelated test failures - i.e. how you had to bump the timeout in this PR

dario-piotrowicz commented:
> Give me some time to think about it, I'm not settled.
> 
> Some thoughts I have for now about `initOpenNextCloudflare()`
> 
> ...

Well there is only one con and many pros, I dislike that we need to then ask users to always call the function, but given the benefits it does seem like it might be worth it

By the way, either way I am not too convinced about the name since this is a dev (`next dev`) only thing and in my opinion  extremely unlikely to be useful outside of dev (because we have other places where we could accept options/tweak things, like in our templates, in the open-next config, etc...) for this reason I would really prefer it if the function had `dev` somewhere in its name (e.g. `initOpenNextCloudflareForDev()`)

vicb commented on .changeset/chilly-dryers-begin.md:
```diff
@@ -0,0 +1,47 @@
+
+import { initOpenNextCloudflareForDev } from "@opennextjs/cloudflare";
+
+/** @type {import('next').NextConfig} */
+const nextConfig = {};
+
+initOpenNextCloudflareForDev();
```
init first ?

```suggestion
initOpenNextCloudflareForDev();

/** @type {import('next').NextConfig} */
const nextConfig = {};
```

vicb commented on .changeset/chilly-dryers-begin.md:
````diff
@@ -0,0 +1,47 @@
+  ```js
+  import { getCloudflareContext } from "@opennextjs/cloudflare";
+
+  const cloudflareContext = await getCloudflareContext();
````
We still should not call this at top level as we are not in a request context

dario-piotrowicz replied:
it's just an example, I didn't want to get too verbose...

would `// ...` lines solve the issue?
or would you prefer a more verbose example?
or maybe I can remove the example altogether since this is also pretty obvious anyways

vicb replied:
I don't like the fact that the example is broken.

As the changeset is already long, I would rather shorten it and remove the example but add details to our documentation.

dario-piotrowicz replied:
example removed 👍

vicb commented on examples/api/e2e/playwright.config.ts:
```diff
@@ -49,5 +49,8 @@ export default defineConfig({
     command: "pnpm preview:worker",
     url: "http://localhost:8770",
     reuseExistingServer: !process.env.CI,
+    // the app uses the `initOpenNextCloudflareForDev` which in CI apparently makes the boot up take slightly longer,
+    // that's why we need a longer timeout here (we just add 10 seconds to the default 60)
+    timeout: !process.env.CI ? undefined : 60 * 1000 + 10 * 1000,
```
nit:

```suggestion
    timeout: 70_000,
```

vicb commented on examples/middleware/next.config.mjs:
```diff
@@ -1,4 +1,8 @@
+import { initOpenNextCloudflareForDev } from "@opennextjs/cloudflare";
+
 /** @type {import('next').NextConfig} */
 const nextConfig = {};
 
+initOpenNextCloudflareForDev();
+
```
```suggestion
initOpenNextCloudflareForDev();

/** @type {import('next').NextConfig} */
const nextConfig = {};

```

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,175 @@
+    const getContextFunctionName = getCloudflareContext.name;
+    const initFunctionName = initOpenNextCloudflareForDev.name;
+    throw new Error(
+      `\n\n\`${getContextFunctionName}\` has been called during development without having used` +
```
```suggestion
      `\n\n\`${getContextFunctionName}\` has been called during development without having called` +
```

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,175 @@
+ * (Currently all the setup that this function performs is making sure that `getCloudflareContext` works as intended,
+ * in the future mode improvements might be added)
+ *
+ * Note: this function should only be called inside the Next.js config file
```
add not that it doesn't need to be `await`ed?

dario-piotrowicz replied:
added 👍

vicb commented on packages/cloudflare/src/api/kvCache.ts:
```diff
@@ -2,7 +2,7 @@ import type { KVNamespace } from "@cloudflare/workers-types";
 import { IgnorableError, RecoverableError } from "@opennextjs/aws/utils/error.js";
 
-import { getCloudflareContext } from "./get-cloudflare-context.js";
+import { getCloudflareContext } from "./cloudflare-context.js";
```
in this file you can remove the following pattern:

```ts
    if (!this.initialized) {
      await this.init();
    }
```

That was only required because `getCloudflareContext` was async

vicb replied:
I'll amend the PR

dario-piotrowicz replied:
ah ok I see what you meant, thanks for the update 👍

vicb reviewed (changes requested):
Cant't wait for this PR to get in 🎉

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,175 @@
+    const getContextFunctionName = getCloudflareContext.name;
+    const initFunctionName = initOpenNextCloudflareForDev.name;
+    throw new Error(
+      `\n\n\`${getContextFunctionName}\` has been called during development without having used` +
```
Could this be a single ```...string...``` to avoid `\n`, `+`?

dario-piotrowicz replied:
we could use string literals but then we'd have to de-dent them, for example using a library like [`ts-dedent`](https://www.npmjs.com/package/ts-dedent) (like wrangler does for example) or by doing that manually (like with a regex replace)

personally I prefer `\n` + `+` (no extra dependency/brittleness to deal with and it is practically as readable), but if you want to use string literals + some de-dent functionality I'm ok with that too

vicb replied:
You could just:

```
      throw new Error(`

\`${getContextFu...
Another line

`);
```

dario-piotrowicz replied:
oh.... I really really dislike that! 🥲

(in my opinion that sort of thing completely breaks files indentation and looks just super messy)

vicb replied:
keep you version then 👍

vicb commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,175 @@
+ * @param cloudflareContext the cloudflare context to add to the node.sj global scope
+ */
+function addCloudflareContextToNodejsGlobal(cloudflareContext: CloudflareContext<CfProperties, Context>) {
+  const global = getInternalGlobalThis<CfProperties, Context>();
+  global[cloudflareContextSymbol] = cloudflareContext;
```
When I see `getInternalGlobalThis(...)` I feel like the function is "doing" something. It is actually a noop and its only purpose it to cast. Can we just directly:

```suggestion
  const global = globalThis as InternalGlobalThis<CfProperties, Context>;
  global[cloudflareContextSymbol] = cloudflareContext;
```

_(global can even be inlined here)_

dario-piotrowicz replied:
yes I was thinking that maybe `getInternalGlobalThis` might be too much as well, I'm happy to remove it 👍

dario-piotrowicz replied:
done 👍

vicb commented on examples/middleware/e2e/playwright.dev.config.ts:
```diff
@@ -0,0 +1,53 @@
+import { defineConfig, devices } from "@playwright/test";
+
+declare var process: { env: Record<string, string> };
```
import { type process } from 'node:process' ?

dario-piotrowicz commented on packages/cloudflare/src/api/cloudflare-context.ts:
```diff
@@ -0,0 +1,175 @@
+};
+
+/**
+ * Gets the globalThis for internal usage
```
done 👍

dario-piotrowicz commented on examples/middleware/e2e/cloudflare-context.spec.ts:
```diff
@@ -0,0 +1,14 @@
+  //       should ideally also be part of the dev cloudflare context
+  //       (this is an upstream wrangler issue: https://github.com/cloudflare/workers-sdk/issues/7812)
+  expect(await cloudflareContextHeaderElement.textContent()).toContain(
+    "keys of `cloudflareContext.env`: MY_VAR, MY_KV"
```
the dependence on the ordering can also easily be fixed by sorting the keys: https://github.com/opennextjs/opennextjs-cloudflare/pull/265/commits/6006cf24cd4d4046213636f08b33183778d9a9a0
(I also had to filter the variables to avoid the ASSETS issue)

I am happy to also add (or instead) an API route and do `array.toContain` whatever you prefer, I'm happy with anything 🙂

vicb replied:
I think API + `array.toContain` would be simpler but the code is already there and working so up to you.

dario-piotrowicz replied:
I just simplified the test there: [6e6d076](https://github.com/opennextjs/opennextjs-cloudflare/pull/265/commits/6e6d076b6071e9d83e904fad20f8e77549f27909)

All we care about here is checking that we can get access to the cloudflare context in the middleware, regarding the context variables we already test that they are present here: https://github.com/opennextjs/opennextjs-cloudflare/blob/a18f826f6b7ece350cadb7e5f9bcc56fdcd84c7f/examples/api/e2e/base.spec.ts#L18-L26

So I think that testing the existence of the `env` object is enough here (and if we wanted to have a comprehensive e2e test for `getCloudflareContext` with all the various variables, `cf`, etc... that should probably live in the api example), what do you think?

vicb commented on .changeset/chilly-dryers-begin.md:
```diff
@@ -0,0 +1,47 @@
+
+introduce new `initOpenNextCloudflareForDev` utility and make `getCloudflareContext` synchronous
+
+introduce a new `initOpenNextCloudflareForDev` function that when called in the [Next.js config file](https://nextjs.org/docs/app/api-reference/config/next-config-js) integrates the Next.js dev server with the open-next Cloudflare adapter. Most noticeably this enables `getCloudflareContext` to work in
```
suggestion: maybe start simple with the TL;DR and add details (or not) in following paragraph.

something like

```
`initOpenNextCloudflareForDev()` must be called from the [Next.js config file](https://nextjs.org/docs/app/api-reference/config/next-config-js) so that your bindings can be accessed  using `getCloudflareContext()` when running the Next.js development server (`next dev`).

`getCloudflareContext` is now sync.

dario-piotrowicz replied:
the changeset title is already a good TL;DR, isn't it?

I think I don't fully understand what you're suggesting, if you want feel free to suggest a code change or to push on the branch

vicb replied:
I think the change set should first highlight what is most important for the users:

1. that the init function must be called from the config
2. that getCloudflareContext is sync

IMO mentioning the edge runtime of our recommendation could be pushed down or moved to the doc site.

dario-piotrowicz replied:
I've updated the changeset I hope the new version works for you

vicb commented:
I've pushed the `kvCache` update

vicb commented:
Could you please update the README and opennext.js.org to tell users to add the init?

vicb commented on examples/create-next-app/e2e/playwright.config.ts:
```diff
@@ -49,8 +49,6 @@ export default defineConfig({
-    // the app uses the `initOpenNextCloudflareForDev` which in CI apparently makes the boot up take slightly longer,
-    // that's why we need a longer timeout here (we just add 10 seconds to the default 60)
-    timeout: !process.env.CI ? undefined : 60 * 1000 + 10 * 1000,
+    timeout: 7_000,
```
oops, should be 70_000... I'm not good at math ;)

vicb commented on examples/api/e2e/playwright.config.ts:
```diff
@@ -1,6 +1,7 @@
 import { defineConfig, devices } from "@playwright/test";
+import nodeProcess from "node:process";
```
Doesn't that work:
```suggestion
import type nodeProcess from "node:process";
```

dario-piotrowicz replied:
no 🤷 

![Screenshot 2025-01-24 at 10 24 22](https://github.com/user-attachments/assets/a5954e63-bd2c-4328-b637-628875905469)

vicb replied:
(the difference being that `import type` does not import anything (i.e. TS only construct)

dario-piotrowicz replied:
updated 👍

vicb commented on .changeset/chilly-dryers-begin.md:
```diff
@@ -0,0 +1,28 @@
+---
+"@opennextjs/cloudflare": patch
```
what about cutting a minor?

It would signal users that there are changes:

- calling the`initOpenNextCloudflareForDev` function,
- `getCloudflareContext` becoming sync

If you do that, please also prepare:
- a PR to update the c3 template to 0.4.x
- a PR to the doc site (copy the current content to 0.3, add the changes in the current content, add a migration from 0.3)

dario-piotrowicz replied:
done 👍

vicb reviewed (approved):
Could you please update https://github.com/opennextjs/opennextjs-cloudflare/blob/main/packages/cloudflare/README.md

Either to ask the user to add the call to the init function
Or if you can update opennext.js.org, remove the content and link to the doc site (I was mostly using the content for the experimental branch when the experimental features were not documented anywhere else)

Great stuff, thanks 🚀

vicb commented on examples/middleware/middleware.ts:
```diff
@@ -1,7 +1,9 @@
-export function middleware(request: NextRequest, event: NextFetchEvent) {
+import { getCloudflareContext } from "@opennextjs/cloudflare";
+
+export async function middleware(request: NextRequest, event: NextFetchEvent) {
```
I don't think this needs to be `async` ?

```suggestion
export function middleware(request: NextRequest, event: NextFetchEvent) {
```

dario-piotrowicz replied:
good spot 👍 , I must have added it before making `getCloudflareContext` sync and forgot to revert it later

vicb commented on packages/cloudflare/README.md:
````diff
@@ -6,65 +6,7 @@ Deploy Next.js apps to Cloudflare!
-
-export default config;
-```
+To get started with the adapter visit the [official get started documentation](https://opennext.js.org/cloudflare/get-started).
````
Can you please also drop the KI section below which is out of date

dario-piotrowicz replied:
done [da78a85](https://github.com/opennextjs/opennextjs-cloudflare/pull/265/commits/da78a853aaadb46feae197ed90c0274d5c4624d9) 👍

vicb reviewed (approved):
Looks great, thanks!

vicb commented:
@dario-piotrowicz, would you mind posting this on Discord once you push a release? (0.4 + getContext sync) 🙏
